File: frontend/Concord/src/api/directCallsService.js
import { request } from './httpClient';
import { buildPagingQuery } from './pagination';

export async function startCall(conversationId, type) {
  return request(`DirectMessages/Conversations/${conversationId}/Calls`, { method: 'POST', body: { Type: type } });
}

export async function acceptCall(conversationId, callId) {
  return request(`DirectMessages/Conversations/${conversationId}/Calls/${callId}:Accept`, { method: 'POST' });
}

export async function declineCall(conversationId, callId) {
  return request(`DirectMessages/Conversations/${conversationId}/Calls/${callId}:Decline`, { method: 'POST' });
}

export async function endCall(conversationId, callId) {
  return request(`DirectMessages/Conversations/${conversationId}/Calls/${callId}:End`, { method: 'POST' });
}

export async function getCalls(conversationId, { page, pageSize } = {}) {
  return request(`DirectMessages/Conversations/${conversationId}/Calls`, { query: buildPagingQuery(page, pageSize) });
}

File: mobile/concord/lib/api/direct_calls_service.dart
import 'api_client.dart';
import 'models/call_response.dart';
import 'models/enums.dart';
import 'models/paged_result.dart';

class DirectCallsService {
  DirectCallsService(this._client);

  final ApiClient _client;

  Future<CallResponse> startCall(String conversationId, CallType type) async {
    final data = await _client.post(
      '/DirectMessages/Conversations/$conversationId/Calls',
      body: {'Type': type.toWire()},
    );
    return CallResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<CallResponse> acceptCall(String conversationId, String callId) async {
    final data = await _client.post('/DirectMessages/Conversations/$conversationId/Calls/$callId:Accept');
    return CallResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<CallResponse> declineCall(String conversationId, String callId) async {
    final data = await _client.post('/DirectMessages/Conversations/$conversationId/Calls/$callId:Decline');
    return CallResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<CallResponse> endCall(String conversationId, String callId) async {
    final data = await _client.post('/DirectMessages/Conversations/$conversationId/Calls/$callId:End');
    return CallResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<PagedResult<CallResponse>> getCalls(
    String conversationId, {
    required int page,
    required int pageSize,
  }) async {
    final data = await _client.get(
      '/DirectMessages/Conversations/$conversationId/Calls',
      query: {'page': page, 'pageSize': pageSize},
    );
    return PagedResult.fromJson(data as Map<String, dynamic>, CallResponse.fromJson);
  }
}
